### ISSUE TITLE
restart from step 10 with external sam files

### ISSUE THREAD

**Issue by bresyd** [jtamames/SqueezeMeta#433]
Hi,

I would like to redo some of my squeezemeta runs including all steps after the mapping using my own set of sam files. 
The rational for this is because I want to do some filtering of the mapped reads based on mapping identity and alignment length (introducing more stringent mapping criteria) and then use these filtered mapped reads to continue the squeezemeta runs.
I already have those filtered sam files and used the identical filenames as the sam files originally produced by squeezemeta.

I was wondering how I would now best go about restarting the runs?
- replace the original sam files in `projectname/data/sam` with my filtered sam files
- `restart.pl projectname -step 10`: but since I don't want to repeat the mapping, I guess I need to hash some of the lines in the mapping script?
- do I need to remove any already existing output files/directories?  

Also, I am now using SQM_1.5 but I would like to do the above also on some of my project created using SQM_1.4. Are there any incompatibility issues I need to be aware of? If possible, I would like to use SQM_1.5 for all my projects since it has concoct as additional binning tool.

Any advice on how to best go about this will be very appreciated
Thanks a lot for your help,
cheers

**Comment by jtamames** [comment]
Hello
I assume you want this for subsequent binning. But take into account that the contig abundance tables that are needed for binning must be redone in order to take into account this new information, and this is done in step 10. So you need to repeat that step. The mapping will not be redone since SqueezeMeta checks for the presence of sam files and if these are present, it does not create them again. But it will generate a new contig coverage table.
So, just copy your sam files and restart from 10.
Best, 
J

**Comment by bresyd** [comment]
Hi,

thanks a lot for your reply. Great to hear that SqueezeMeta checks for sam files and won't repeat the mapping in case they are present. This should make everything fairly straight forward.
Do you see any issues with using version 1.5 for the restart for projects that I initially ran using version 1.4?

cheers

**Comment by jtamames** [comment]
Hello
It will recognize that your project is 1.4 and will ask you to run `versionchange.pl` to transform them as needed for 1.5. Then it should work fine.
Best,
J

**Comment by bresyd** [comment]
Hi,

I did as we outlined above but ran into an issue.
After running `versionchange.pl` I manually included the path to my SqueezeMeta_1.5 conda installation in the SqueezeMeta_conf.pl file;
Then I restarted the run at step 10 using `restart.pl msm89_coass2_clustered -step 10`
It then finished steps 10, 11, 12 and 13 but ran into an issue at the binning step:
```2 hours, 59 minutes, 41 seconds]: STEP14 -> BINNING: 14.runbinning.pl
WARNING in STEP14 -> No binner found for concoct
WARNING in STEP14 -> No binner found for maxbin
WARNING in STEP14 -> No binner found for metabat2
 ```

I checked the SqueezeMeta_conf.pl file and manually added the following lines (which I took from one of my previous runs using the 1.5 version):
```
%bindirs = ("metabat2","$resultpath/metabat2","maxbin","$resultpath/maxbin");
%dasdir      = ("DASTool","$resultpath/DAS/$projectname\_DASTool\_bins");          #-- Directory for DASTool resu$
$concoct_dir        = "$installpath/bin/CONCOCT-1.1.0";
%binscripts         = ('maxbin',"$installpath/lib/SqueezeMeta/bin_maxbin.pl",'metabat2',"$installpath/lib/SqueezeMeta/bin_metabat2.pl",'concoct',"$installpath/lib/SqueezeMeta/bin_concoct.pl");
```
I also changed the numbering of the parameters from step 14 onwards in the parameters.pl file.

Then I restarted at step 14 `restart.pl msm89_coass2_clustered -step 14` and got the following error
```
[0 seconds]: STEP14 -> BINNING: 14.runbinning.pl
  Running concoct from /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_concoct.pl
mkdir: cannot create directory ‘/scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct’: No such file or directory
sh: 1: cannot create /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct/coverage_table.tsv: Directory nonexistent
sh: 1: cannot create /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct/concoct_int/clustering_merged.csv: Directory nonexistent
```
I couldn't really figure out what the issue was, but then I created the intermediate/binners/concoct directory manually and restarted again, now getting the following error:
```
[0 seconds]: STEP14 -> BINNING: 14.runbinning.pl
  Running concoct from /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_concoct.pl
[E::hts_open_format] Failed to open file /scratch/btschits/sqm_1.5/msm89_coass2_clustered/temp/*.sorted.bam
ERROR: fail to open index BAM file '/scratch/btschits/sqm_1.5/msm89_coass2_clustered/temp/*.sorted.bam'
Traceback (most recent call last):
  File "/home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/bin/CONCOCT-1.1.0/scripts/concoct_coverage_table.py", line 91, in <module>
    generate_input_table(args.bedfile, args.bamfiles, samplenames=samplenames)
  File "/home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/bin/CONCOCT-1.1.0/scripts/concoct_coverage_table.py", line 41, in generate_input_table
    sys.stderr.write(out)
```
I checked the content of the temp dir:
```
ls -ltr temp/ | tail -10
-rw-r--r-- 1 btschits biogeo    2560418 Feb 22 18:18 fun3tax_40.noidfilter.wranks
-rw-r--r-- 1 btschits biogeo   12576279 Feb 22 18:18 09.msm89_coass2_clustered.allorfs
-rw-r--r-- 1 btschits biogeo   14082726 Feb 22 18:18 09.msm89_coass2_clustered.allorfs.noidfilter
-rw-r--r-- 1 btschits biogeo 1313001277 Feb 22 18:18 bincontigs.fasta
-rw-r--r-- 1 btschits biogeo        322 Feb 22 18:18 stats.txt
drwxr-xr-x 7 btschits biogeo         14 Feb 23 20:21 ..
drwxr-xr-x 2 btschits biogeo        145 Feb 23 20:24 .
-rw-r--r-- 1 btschits biogeo    8054319 Feb 23 20:24 msm89_coass2_clustered.contigs.bed
-rw-r--r-- 1 btschits biogeo 1406080497 Feb 23 20:24 msm89_coass2_clustered.choppedcontigs.fasta
-rw-r--r-- 1 btschits biogeo       3354 Feb 23 20:24 msm89_coass2_clustered.log

```
so it never created the bam files required for the concoct binning.

Here are some of the lines in the syslog file with the commands for sorting and bam conversion
```
Restarting project msm89_coass2_clustered, Wed Feb 23 20:24:22 2022
[0 seconds]: STEP14 -> 14.runbinning.pl

RUNNING concoct
  Working for msm89_coass2_clustered.5146_C.sam
  Creating sorted BAM for msm89_coass2_clustered.5146_C.sam:  sort /scratch/btschits/sqm_1.5/msm89_coass2_clustered/data/sam/msm89_coass2_clustered.5146_C.sam -o /scratch/btschits/sqm_1.5/msm89_coass2_clustered/temp/msm89_coass2_clustered.5146_C.sorted.bam -@40 > /dev/null 2>&1
  Indexing sorted BAM:  index /scratch/btschits/sqm_1.5/msm89_coass2_clustered/temp/msm89_coass2_clustered.5146_C.sorted.bam > /dev/null 2>&1
```
and this is how far I got with this.
Do you have any idea how to maybe solve this issue?

Thanks again for your help.

**Comment by bresyd** [comment]
Hi,
I noticed another line (for samtools) that was missing from the SqueezeMeta_conf.pl file and again added it manually
`$samtools_soft      = "LD_LIBRARY_PATH=\$LD_LIBRARY_PATH:$installpath/lib $installpath/bin/samtools";`

Then I restarted again at step 14 (after deleting the previously created ` intermediate/binners/concoct` dir), here is the output from the terminal:
```
restart.pl msm89_coass2_clustered -step 14
[0 seconds]: STEP14 -> BINNING: 14.runbinning.pl
  Running concoct from /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_concoct.pl
mkdir: cannot create directory ‘/scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct’: No such file or directory
sh: 1: cannot create /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct/coverage_table.tsv: Directory nonexistent
sh: 1: cannot create /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct/concoct_int/clustering_merged.csv: Directory nonexistent
Cannot open bin directory /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct
ERROR in STEP14 -> /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_concoct.pl
mkdir: cannot create directory ‘/scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct’: No such file or directory
Can't open /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct directory
ERROR in STEP14 -> 14.runbinning.pl
[13 minutes, 0 seconds]: STEP15 -> DAS_TOOL MERGING: 15.dastool.pl
mkdir: cannot create directory ‘/scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/DAS’: No such file or directory
Can't open /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct directory
ERROR in STEP15-> 15.dastool.pl
Can't open  directory
```
However, now it did create the sorted bam files in the temp dir:
```
ls -ltr temp/ |tail | awk {'print $5,$9'}  
984993948 msm89_coass2_clustered.sample_N.sorted.bam
10107784 msm89_coass2_clustered.sample_N.sorted.bam.bai
1810655474 msm89_coass2_clustered.sample_G.sorted.bam
8688896 msm89_coass2_clustered.sample_G.sorted.bam.bai
234595021 msm89_coass2_clustered.5146_I.sorted.bam
193 .
5996360 msm89_coass2_clustered.5146_I.sorted.bam.bai
8054319 msm89_coass2_clustered.contigs.bed
1406080497 msm89_coass2_clustered.choppedcontigs.fasta
3579 msm89_coass2_clustered.log

``` 

So there is still something wrong with the concoct binning script I guess;

**Comment by fpusan** [comment]
`mkdir: cannot create directory ‘/scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/concoct’: No such file or directory`
Maybe you need to create the `binners` directory manually before the `concoct` directory can be created?

**Comment by bresyd** [comment]
Hi,

thanks for the  advice. I restarted again after manually creating the `intermediate/binners/concoct` dir and concoct did run successfully!
However, I than ran into an error with maxbin, and later on DAStool (while metabat2 finished successfully)
```
Running maxbin from /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_maxbin.pl
Error running command:    perl /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/bin/MaxBin/run_MaxBin.pl -thread 40 -contig /scratch/btschits/sqm_1.5/msm89_coass2_clustered/temp/bincontigs.fasta -abund_list /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/maxbin/abund.list -out /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/maxbin/maxbin -markerpath /scratch/btschits/sqm_1.4/db/marker.hmm at /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_maxbin.pl line 132.
ERROR in STEP14 -> /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_maxbin.pl
wc: /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/maxbin/: Is a directory
WARNING in STEP14 -> /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_maxbin.pl. No maxbin results!
  Running metabat2 from /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib/SqueezeMeta/bin_metabat2.pl
[1 hours, 9 minutes, 0 seconds]: STEP15 -> DAS_TOOL MERGING: 15.dastool.pl
mkdir: missing operand
Try 'mkdir --help' for more information.
unzip:  cannot find or open /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/bin/DAS_Tool/db.zip, /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/bin/DAS_Tool/db.zip.zip or /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/bin/DAS_Tool/db.zip.ZIP.
Error running command:    LD_LIBRARY_PATH=/home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/lib PATH=/home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/bin:$PATH /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/bin/DAS_Tool/DAS_Tool -i /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/DAS/concoct.table,/scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/DAS/metabat2.table -l concoct,metabat2 -c /scratch/btschits/sqm_1.5/msm89_coass2_clustered/results/01.msm89_coass2_clustered.fasta --write_bins 1 --score_threshold 0 --search_engine diamond -t 40 -o /scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/DAS/msm89_coass2_clustered --db_directory /scratch/btschits/sqm_1.4/db at /home/btschits/miniconda3/envs/SqueezeMeta_1.5.0_deb10/SqueezeMeta/scripts/15.dastool.pl line 93.
mv: missing destination file operand after '/scratch/btschits/sqm_1.5/msm89_coass2_clustered/intermediate/binners/DAS/msm89_coass2_clustered_DASTool_bins/*'
Try 'mv --help' for more information.
Can't open  directory
```
Any ideas and sorry for taking up so much of your time for this issue.

**Comment by jtamames** [comment]
Hello
This is pretty wierd. The script `versionchange.pl `must do all these changes. It looks like it didn't work properly. How did you run it? What messages did you get?
Best,
J

**Comment by bresyd** [comment]
Hi,

thanks for getting back to me again. 

I just noticed something which might be related to the issue. Before I ran `versionchangle.pl projectname` , I actually removed all files related to steps 10 or later, including all files in the `results `directory starting from 10.*, and also all bin directories from the `results `dir. I did this because I thought I won't need them since I am going to restart at step 10, but maybe the versionchange script would have actually needed them, otherwise it wouldn't know anything about the binning tools I used (maybe that is why I had to manually copy some lines into the `SqueezeMeta_conf.pl` file after the conversion, see above). So I am going to repeat everything without deleting any files before running` versionchange.pl `and see if things work out then; I will let you know how that goes.

And just to make sure that the issue is not related to my filtered sam files, I repeated the SQM run from scratch, using version 1.5 and copying my filtered sam files into the `data/sam` directory right after the run started; that run finished without any issue including the binning with concoct, maxbin and metabat2 plus DAStool.

cheers
